### PR TITLE
Lksm

### PR DESCRIPTION
**Pull request by qwer951123**

### CODE REVIEW

**Comment by qwer951123** [comment]
@shengda

File: rewards/run_lksm.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import { generateMerkle } from "./generate_merkle";
+import { getAccounts } from "./query-accounts";
+import { getLKSMRawBalance, getLKSMBalance } from "./query-balance-lksm";
+import { getTaiKsmBalance, getTaiKsmRawBalance } from "./query-balance-taiksm";
+import { getClaimedLKSMAccounts } from "./query-claimed-lksm-accounts";
+import { submitMerkle } from "./submit-merkle";
+
+const BLOCK = 2372400;
+
+const main = async () => {
```
Comment (shengda) [inline rewards/run_lksm.ts:11]: Do you want to run LKSM and taiKSM as separate pipeline or a single one?

File: rewards/run_taiksm.ts [hunk 2]
```diff
@@ -17,13 +20,22 @@ const main = async () => {
...
+    await getClaimedLKSMAccounts(block);
     await getTaiKsmRawBalance(block);
     await getTaiKsmBalance(block);
+    await getLKSMRawBalance(block);
+    await getLKSMBalance(block);
+    // 1. calculate tai reward
+    // 2. calculate lksm reward in taiKSM
     await distributeTaiKsm(block);
+    // calculate lksm reward in free
+    await distributeLKSM(block);
```
Comment (shengda) [inline rewards/run_taiksm.ts:32]: If you want to run lksm distribution as a separate pipeline, then this task is not part of the taiKSM pipeline.

File: rewards/distribute-lksm.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+            // write reward records to file
+
+            let fd = await fs.promises.open(distributionFile, "w");
+            let distributionContent = 'AccountId, 0x0000000000000000000100000000000000000080\n';
+
+            for (const address in accountBalance) {
+                if (!address)   continue;
+                distributionContent += `${address},${incressRewards[address].toString()}\n`;
+                distributionContent += `reserved-${address},${reserved[address].toString()}\n`;
```
Comment (shengda) [inline rewards/distribute-lksm.ts:147]: The format is: [user address], reserve-[token address]

**Review by shengda — CHANGES_REQUESTED** [review]

**Pull request merged**